### PR TITLE
5072 - updating bn 1.x

### PR DESCRIPTION
**Pull request by luu-alex**
## Description
#5072
Please include a summary of the changes and be sure to follow our [Contribution Guidelines](../blob/1.x/CONTRIBUTIONS.md).

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` with success.
- [ ] I have tested the built `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

### CODE REVIEW

**Review by jdevcs — COMMENTED** [review]
We will also need to update https://github.com/ChainSafe/web3.js/blob/2c0324af2da467ee1acb72452f20000e916e4306/packages/web3-utils/src/utils.js#L273 and might need to remove `number-to-bn` dependency. also update ethers to 5.6.8

**Comment by luu-alex** [comment]
@jdevcs  the dependancy `number-to-bn` uses `bigNumber.js` so that should be fine? will update the others

**Comment by luu-alex** [comment]
The failing testcase that is commented is related to this issue #5071. will need to investigate further

File: test/eth.accounts.signTransaction.js [hunk 1]
```diff
@@ -652,29 +652,29 @@ var tests = [
...
-            data: "",
-            common: commonLondon
-        },
-        // signature from eth_signTransaction
-        rawTransaction: "0x02f86e018084c733124884cb72ec20826a4094f0109fc8df283027b6285cc889f5aa624eac1f55843b9aca0080c001a08896fb9a5c033e0163b073cf7a951a1db2dca41b26b4188f13a05158eb26fd32a005e8855691199cd0b6dcae88f3325c374e2f0697b9c528a5c10d5bd8dfb6a3e3",
-        oldSignature: "0x02f86e018084c733124884cb72ec20826a4094f0109fc8df283027b6285cc889f5aa624eac1f55843b9aca0080c001a08896fb9a5c033e0163b073cf7a951a1db2dca41b26b4188f13a05158eb26fd32a005e8855691199cd0b6dcae88f3325c374e2f0697b9c528a5c10d5bd8dfb6a3e3",
-        transactionHash: "0xd5b7290a477b9c421d39e61d0f566ec33276fb49b9ff85cfd6152a18f1c92dab",
-        messageHash: "0x17e20e530a889ce52057de228b5b97edcad6002468d723346cd0b6b7a9943457"
-    },
+    // {
```
Comment (avkos) [inline test/eth.accounts.signTransaction.js]: why the test is commented on?

**Review by avkos — CHANGES_REQUESTED** [review]

File: test/eth.accounts.signTransaction.js (hunk 1 above)
Comment (luu-alex) [reply test/eth.accounts.signTransaction.js]: This testcase is not correct, we were using bn version 4.11.9 and an error was [found](https://github.com/indutny/bn.js/issues/151) parsing hexstirngs incorrectly. it has been incorrectly converting hexstrings to decimals. 
```
> var a = new web3.utils.BN("0x3B9ACA00", 16)
undefined
> a
BN { negative: 0, words: [ 60475904, 526 ], length: 2, red: null }
> 
```
But `0x3B9ACA00` should be `1000000000`

So with the new bn.js update for our dependancies, there was a [breaking change](https://github.com/indutny/bn.js/releases/tag/v5.0.0) and now this testcase is failing and i had to comment it out. Will need to investigate this

**Review by nikoulai — APPROVED** [review]

File: test/eth.accounts.signTransaction.js (hunk 1 above)
Comment (nazarhussain) [reply test/eth.accounts.signTransaction.js]: I believe it's a very important test in `signTransaction` so we should not disable it, but rather investigate it and also include its fix in the same release. 

If the upgrade of `bn.js` is causing some misbehavior to transaction signature then we should report the issue to them and don't upgrade `bn.js` right away.

File: package.json [hunk 2]
```diff
@@ -101,9 +101,10 @@
         "@types/bignumber.js": "^4.0.2",
         "@types/bn.js": "^4.11.6",
         "@types/node": "^12.12.68",
+        "@types/prettier": "2.6.0",
         "babel-loader": "^8.1.0",
         "bignumber.js": "^9.0.1",
-        "bn.js": "^4.11.9",
+        "bn.js": "^5.2.1",
```
Comment (nazarhussain) [inline package.json:107]: We need to update its type as well. 

https://github.com/ChainSafe/web3.js/blob/2c0324af2da467ee1acb72452f20000e916e4306/package.json#L102

**Review by nazarhussain — CHANGES_REQUESTED** [review]

File: test/eth.accounts.signTransaction.js (hunk 1 above)
Comment (nikoulai) [reply test/eth.accounts.signTransaction.js]: From the tests I have run, the misbehavior seems to be only in the 4.11.9 version, so I just updated the hex values in the test and reenabled it.

I noticed that ethereumjs-utils uses bn:4.11.6, maybe, we should let them know

**Review by nikoulai — COMMENTED** [review]
Awesome!

In 4.x I saw something like that for the handling negative hex:

```
 if (typeof value == "string" && value.includes("0x")) {
        const [negative, hexValue] = value.toLocaleLowerCase().startsWith('-') ? ["-", value.slice(3)] : ["", value.slice(2)];
        return new BN(negative + hexValue, 16);
    }
    else {
        return new BN(value);
    }
```

What would you say to refactor?

File: test/eth.accounts.signTransaction.js (hunk 1 above)
Comment (luu-alex) [reply test/eth.accounts.signTransaction.js]: test is uncommented and works fine now with @nikoulai's [fix](https://github.com/ChainSafe/web3.js/pull/5074/files#diff-bd97e9c608060469030c1d6a655baa533c0aa0b11939b0670ebdb41aa51d786bR663), the issue is `bn` errors when given strings starting with`0x`, we shouldn't have passed ` bn('0x3B9ACA00')` but instead it should have been `bn('cb72ec20', 16),`

File: packages/web3-utils/src/utils.js [hunk 3]
```diff
@@ -534,8 +534,29 @@ var toNumber = function(value) {
     return typeof value === 'number' ? value : hexToNumber(toHex(value));
 }
 
+// 1.x currently accepts 0x... strings, bn.js after update doesn't. it would be a breaking change
```
Comment (luu-alex) [inline packages/web3-utils/src/utils.js:537]: Thinking about this now we should probably remove this, I dont think we should be changing how BN is being created... the [test case](https://github.com/ChainSafe/web3.js/pull/5074/files#diff-bd97e9c608060469030c1d6a655baa533c0aa0b11939b0670ebdb41aa51d786bR663) is working fine and this is adding an extra feature this PR should just be focusing on upgrading and fixing `BN`.

Comment (nikoulai) [reply packages/web3-utils/src/utils.js:537]: You are right about the scope, but, shouldn't we handle the <change> introduced too? (you mean to do it in another PR? )

**Comment by luu-alex** [comment]
Added an extra [test](https://github.com/ChainSafe/web3.js/pull/5074/files#diff-bd97e9c608060469030c1d6a655baa533c0aa0b11939b0670ebdb41aa51d786bR702), works properly as in [4.x](https://github.com/ChainSafe/web3.js/pull/5070/files#diff-285b5b91489566a7887cb8efb734e5abc35f3b2f1d42d960b077b44d9f48bef6R834) as well.

**Review by nazarhussain — APPROVED** [review]

**Review by jdevcs — APPROVED** [review]

**Review by avkos — APPROVED** [review]

**Comment by wbt** [comment]
FYI, this turned out to be a breaking change on nonbreaking semver, as discussed in more detail in the above-linked comment, and possibly foreshadowed [here](https://github.com/ChainSafe/web3.js/pull/5132).

**Comment by arijoon** [comment]
As @wbt noted, this was a breaking change. When exposing deps directly (such as `BN`), please bump major version on web3 if any of these deps have a major upgrade

**Pull request merged**